### PR TITLE
Spin loop pause function redux

### PR DESCRIPTION
**Pull request by mstewartgallus**
GitHub's interface is screwy.

This is the same PR as https://github.com/rust-lang/rust/pull/40537

### CODE REVIEW

**Comment by BurntSushi** [comment]
@alexcrichton What do you think? I saw you raised some concerns [here](https://github.com/rust-lang/rust/pull/40537#issuecomment-292972023), but I think they've been addressed in this PR.

**Comment by alexcrichton** [comment]
Yes the docs seems reasonable, although I'm not a huge fan of the name `hint_core_should_pause`. I don't have many suggestions, though. Is there precedent in, say, C++ for this?

**Comment by sstewartgallus** [comment]
@alexcrichton There is no precedent in C or C++ but there is in Java see http://openjdk.java.net/jeps/285 .

**Comment by alexcrichton** [comment]
Thanks for the pointer! Looks like Java called this `java.lang.Thread.onSpinWait` (FWIW)

Not sure if that means much for us...

**Comment by clarcharr** [comment]
The [`spin`](https://github.com/mvdnes/spin-rs/blob/master/src/util.rs) crate calls this function `cpu_relax`; maybe we should do something similar?

**Comment by alexcrichton** [comment]
Ok this seems like a reasonable API to me to have in libstd, and we've probably had enough bikeshedding at this point! The name seems totally reasonable to me in terms of it's still unstable and we'll have a final pass before stabilizing regardless.

@sstewartgallus if you'd like to rebase I'll r+!

**Comment by alexcrichton** [comment]
Ah sorry looks like I miseed that rebase before bors caught it, wanna try again?

**Comment by alexcrichton** [comment]
@sstewartgallus looks like tidy is failing, and mind squashing this into one commit?

File: src/doc/unstable-book/src/hint-core-should-pause.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# `hint_core_should_pause`
```
Comment (mzji) [inline src/doc/unstable-book/src/hint-core-should-pause.md]: This file should be moved to the ` 	library-features` directory.

File: src/doc/unstable-book/src/SUMMARY.md [hunk 2]
```diff
@@ -221,3 +222,4 @@
     - [windows_handle](library-features/windows-handle.md)
     - [windows_net](library-features/windows-net.md)
     - [windows_stdio](library-features/windows-stdio.md)
+    - [zero_one](library-features/zero-one.md)
```
Comment (mzji) [inline src/doc/unstable-book/src/SUMMARY.md]: There isn't `zero_one` feature anymore, is this a typo?

File: src/doc/unstable-book/src/SUMMARY.md [hunk 3]
```diff
@@ -145,6 +145,7 @@
     - [future_atomic_orderings](library-features/future-atomic-orderings.md)
     - [get_type_id](library-features/get-type-id.md)
     - [heap_api](library-features/heap-api.md)
+    - [hint_core_should_pause](hint-core-should-pause.md)
```
Comment (mzji) [inline src/doc/unstable-book/src/SUMMARY.md]: Please change the link to `library-features/hint-core-should-pause.md` .

**Pull request merged**